extra/mariabackup/xtrabackup.cc

```diff
@@ -5106,6 +5162,21 @@ class BackupStages {
 				return false;
 			}
 
+			// Copy just enough log to cover the latest commit.
+			// Meanwhile, there could be some active transactions
+			// that are modifying the database and writing more
+			// log. Not copying log for them will save busy work
+			// and avoid some rollback of the incomplete
+			// transactions after the backup has ben restored.
+			// We find the current InnoDB LSN by executing
+			// SHOW ENGINE INNODB STATUS, which would be written
+			// to mysql.general_log, which could be an
+			// ENGINE=CSV table that we just copied
+			// above.
+			if (!backup_wait_for_commit_lsn()) {
+				return false;
+			}
```

Comment:
Currently we execute the following:
```sql
BACKUP STAGE BLOCK_COMMIT;
# copy mysql/general_log.*
SHOW ENGINE INNODB STATUS;
```
I can repeat an occasional failure of the test `mariabackup.log_tables`. Accessing `general_log` would still fail because `general_log.CSM` is marked as "crashed". That marking happened during the execution of `BACKUP STAGE BLOCK_COMMIT`:
```
#0  write_meta_file (meta_file=21, rows=107, dirty=false) at /mariadb/10.11/storage/csv/ha_tina.cc:364
#1  0x00005557dde57f8a in free_share (share=0x7f5de007bde8) at /mariadb/10.11/storage/csv/ha_tina.cc:449
#2  0x00005557dde581c2 in ha_tina::close (this=0x7f5de00764c0) at /mariadb/10.11/storage/csv/ha_tina.cc:1017
#3  0x00005557ddc5149e in handler::ha_close (this=0x7f5de00764c0) at /mariadb/10.11/sql/handler.cc:3517
#4  0x00005557ddad6663 in closefrm (table=table@entry=0x7f5de00746e8) at /mariadb/10.11/sql/table.cc:4649
#5  0x00005557ddbcabdc in intern_close_table (table=0x7f5de00746e8) at /mariadb/10.11/sql/table_cache.cc:230
#6  0x00005557ddbcb545 in tc_purge () at /mariadb/10.11/sql/table_cache.cc:324
#7  0x00005557dd97e8e2 in purge_tables () at /mariadb/10.11/sql/sql_base.cc:331
#8  0x00005557dd97e962 in flush_tables (thd=thd@entry=0x7f5de4002758, flag=flag@entry=FLUSH_SYS_TABLES) at /mariadb/10.11/sql/sql_base.cc:561
#9  0x00005557ddbd4292 in backup_block_commit (thd=thd@entry=0x7f5de4002758) at /mariadb/10.11/sql/backup.cc:404
#10 0x00005557ddbd44f3 in run_backup_stage (thd=thd@entry=0x7f5de4002758, stage=BACKUP_LOCK_COMMIT) at /mariadb/10.11/sql/backup.cc:129
#11 0x00005557dd9eebe0 in mysql_execute_command (thd=thd@entry=0x7f5de4002758, is_called_from_prepared_stmt=is_called_from_prepared_stmt@entry=false) at /mariadb/10.11/sql/sql_parse.cc:5241
#12 0x00005557dd9f1e97 in mysql_parse (thd=thd@entry=0x7f5de4002758, rawbuf=<optimized out>, length=<optimized out>, parser_state=parser_state@entry=0x7f5e08132410) at /mariadb/10.11/sql/sql_parse.cc:8161
#13 0x00005557dd9f3393 in dispatch_command (command=command@entry=COM_QUERY, thd=thd@entry=0x7f5de4002758, packet=packet@entry=0x7f5de400d3f9 "BACKUP STAGE BLOCK_COMMIT", packet_length=packet_length@entry=25, blocking=blocking@entry=true) at /mariadb/10.11/sql/sql_parse.cc:1895
```
So, there clearly is a race condition in `mariadb-backup --backup`: we may or may not copy the `mysql.general_log` at the right moment.

@montywi 0efe5dc7e15ee5e54661ba9dea40731baae207e7 would work around this issue. Is that acceptable, or should the write of `BACKUP STAGE BLOCK_COMMIT` to `mysql.general_log` be suppressed somehow?

Reply:
I took a closer look at this today. Actually there is no problem with the `BACKUP STAGE BLOCK_COMMIT` itself (sorry, when I posted the stack trace yesterday, I should have paid attention to the `dirty=false`). The problem is the log tables are being copied asynchronously, and this copying is not guaranteed to finish before the following is executed:
```c++
			if (!m_common_backup.close_log_tables()) {
				msg("Error on close log tables");
				return false;
			}
```
I downloaded a `var.tar.gz` of a failure from buildbot.mariadb.org, and in it I found the statement `SHOW ENGINE INNODB STATUS` at the end of the file `mysql-test/var/log/mariabackup.log_tables/mysqld.1/data/mysql/general_log.CSV`:
```
"2024-09-04 06:07:02.516380","root[root] @ localhost []",6,1,"Query","BACKUP STAGE BLOCK_COMMIT"
"2024-09-04 06:07:02.517424","root[root] @ localhost []",5,1,"Query","SHOW ENGINE INNODB STATUS"
```
After I revised the logic so that  `m_common_backup.close_log_tables()` will be invoked before submitting further SQL statements, the test would no longer fail for me.

Reply:
2785816cd429fb3b907c70b88c6a264dcb256bed fixes this. I filed [MDEV-34871](https://jira.mariadb.org/browse/MDEV-34871) for letting the `m_common_backup.close_log_tables()` remain where it used to be.